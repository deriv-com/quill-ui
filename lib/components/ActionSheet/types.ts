import { ComponentPropsWithoutRef } from "react";
import { ExcludeAllNull, TLeftOrRight } from "../../types";

interface OpenType {
    isOpen?: boolean | undefined;
    onOpen?: () => void;
}

interface actionSheetRootCVA {
    position?: TLeftOrRight;
    show?: boolean;
}

interface actionSheetFooterCVA {
    alignment?: "vertical" | "horizontal";
}

export interface RootProps
    extends ComponentPropsWithoutRef<"div">,
        ExcludeAllNull<actionSheetRootCVA>,
        OpenType {
    onClose?: () => void;
    type?: "modal" | "non-modal";
    expandable?: boolean;
}

export type RootPosition = RootProps["position"];

interface ActionType {
    content: string;
    onAction: () => void;
}

export interface FooterProps
    extends ComponentPropsWithoutRef<"div">,
        actionSheetFooterCVA {
    primaryAction?: ActionType;
    secondaryAction?: ActionType;
    shouldCloseOnPrimaryButtonClick?: boolean;
    shouldCloseOnSecondaryButtonClick?: boolean;
}

export type FooterAlignment = FooterProps["alignment"];
