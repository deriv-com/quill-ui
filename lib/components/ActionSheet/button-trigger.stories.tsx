import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import { ActionSheetExample } from "./mocks/example";
import {
    LabelPairedPlaceholderCaptionBoldIcon,
    StandaloneXmarkRegularIcon,
} from "@deriv/quill-icons";

const meta: Meta = {
    title: "Components/Action Sheet/Button Trigger",
    component: ActionSheetExample,
    tags: ["autodocs"],
    args: {
        onClose: fn(),
        title: "Title",
        description: "Description",
        closeIcon: <StandaloneXmarkRegularIcon />,
        icon: <LabelPairedPlaceholderCaptionBoldIcon />,
    },
    argTypes: {
        show: { table: { disable: true } },
        handleOpen: { table: { disable: true } },
        handleClose: { table: { disable: true } },
        isOpen: {
            description:
                "If you wish to manage the opening and closing states, transmit the `open` state from your component. Set the initial value to `false` when passing it.",
        },
        onOpen: {
            description:
                "Pass your callback function using this method. It will be triggered on the open function. If you are passing the `isOpen` state, provide your `open` `setState` function like this: `onOpen={() => setIsOpen(true)}`.",
        },
        onClose: {
            description:
                "Pass your callback function using this method. It will be triggered on the open function.",
        },
        expandable: {
            control: { type: "boolean" },
            description:
                "This prop controls the expandability of the bottom sheet.",
        },
        type: {
            options: ["modal", "non-modal"],
            control: { type: "radio" },
            description:
                "This property is used to specify the type, which offers two options: `modal` and `non-modal`. When set to `modal`, it adds an overlay over the entire body, and clicking on it will close the action sheet. On the other hand, when set to `non-modal`, the area outside of the action sheet remains interactive and no overlay.",
        },
        position: {
            options: ["left", "right"],
            control: { type: "radio" },
            description: "This prop will make bottom sheet expandable",
        },
        title: {
            description: "Title for `ActionSheet.Header`",
        },
        description: {
            description: "Description for `ActionSheet.Header`",
        },
        closeIcon: {
            description:
                "This props allowed you to pass in icon for close button in `ActionSheet.Header`",
        },
        icon: {
            description:
                "This props allowed you to pass in icon for `ActionSheet.Header`",
        },
        primaryAction: {
            control: {
                type: "object",
            },
            description:
                "This prop is meant for `ActionSheet.Footer`. It accepts two property: `Content`, which accepts a string, and `onAction`, which takes a function.",
        },
        secondaryAction: {
            control: {
                type: "object",
            },
            description: "Same as `primaryAction`",
        },
        alignment: {
            control: {
                type: "radio",
            },
            options: ["vertical", "horizontal"],
            description:
                "This prop is for `ActionSheet.Footer` buttons alignment",
        },
    },
} satisfies Meta<typeof ActionSheetExample>;

export default meta;
type Story = StoryObj<typeof ActionSheetExample>;

export const ButtonTrigger: Story = {
    args: {
        expandable: true,
        type: "modal",
        position: "right",
        primaryAction: {
            content: "Primary",
            onAction: () => null,
        },
        secondaryAction: {
            content: "Secondary",
            onAction: () => null,
        },
        alignment: "vertical",
    },
};
