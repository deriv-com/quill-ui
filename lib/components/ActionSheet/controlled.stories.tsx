import type { Meta, StoryObj } from "@storybook/react";

import { ActionSheetExampleControlled } from "./mocks/example";
import { fn } from "@storybook/test";
import {
    LabelPairedPlaceholderCaptionBoldIcon,
    StandaloneXmarkRegularIcon,
} from "@deriv/quill-icons";

const icons: Record<string, object | null> = {
    with_icon: <LabelPairedPlaceholderCaptionBoldIcon />,
    none: null,
};

const meta: Meta = {
    title: "Components/Action Sheet/Controlled",
    component: ActionSheetExampleControlled,
    tags: ["autodocs"],
    parameters: {
        controls: { hideNoControlsWarning: true },
    },
    args: {
        isOpen: false,
        onClose: fn(),
        title: "Title",
        description: "Description",
        type: "non-modal",
        closeIcon: <StandaloneXmarkRegularIcon />,
        onOpen: fn(),
        isPrimaryButtonDisabled: false,
        isSecondaryButtonDisabled: false,
        shouldCloseOnDrag: true,
        shouldCloseOnPrimaryButtonClick: true,
        shouldCloseOnSecondaryButtonClick: true,
        fullHeightOnOpen: false,
    },
    argTypes: {
        isOpen: {
            description:
                "If you wish to manage the opening and closing states, transmit the `open` state from your component. Set the initial value to `false` when passing it.",
            control: { type: "boolean" },
        },
        isPrimaryButtonDisabled: {
            control: { type: "boolean" },
            description:
                "This prop controls if primary button is disabled or not.",
        },
        isSecondaryButtonDisabled: {
            control: { type: "boolean" },
            description:
                "This prop controls if secondary button is disabled or not.",
        },
        show: { table: { disable: true } },
        shouldCloseOnDrag: {
            control: { type: "boolean" },
            description:
                "This prop controls if action sheet should be closed on drag down or not. Default value: false. Property should be passed to ActionSheet.Portal",
        },
        shouldDetectSwipingOnContainer: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            control: { type: "boolean" },
            description:
                "This prop controls if swiping will be detected on the whole Action Sheet, not only on Handlebar. Default value: false. Property should be passed to ActionSheet.Portal",
        },
        showHandlebar: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            control: { type: "boolean" },
            description:
                "This prop controls if Handlebar should be visible or not. Default value: true",
        },
        fullHeightOnOpen: {
            description:
                "This props controls is full height when open the actionsheet",
            control: { type: "boolean" },
            table: { type: { summary: "boolean" } },
        },
        handleOpen: { table: { disable: true } },
        handleClose: { table: { disable: true } },
        onOpen: {
            description:
                "Pass your callback function using this method. It will be triggered on the open function.",
        },
        onClose: {
            description:
                "Pass your callback function using this method. It will be triggered on the open function.",
        },
        expandable: {
            control: { type: "boolean" },
            description:
                "This prop controls the expandability of the bottom sheet.",
        },
        type: {
            options: ["modal", "non-modal"],
            control: { type: "radio" },
            description:
                "This property is used to specify the type, which offers two options: `modal` and `non-modal`. When set to `modal`, it adds an overlay over the entire body, and clicking on it will close the action sheet. On the other hand, when set to `non-modal`, the area outside of the action sheet remains interactive and no overlay.",
        },
        position: {
            options: ["left", "right"],
            control: { type: "radio" },
            description: "This prop will make bottom sheet expandable",
        },
        icon: {
            description:
                "This props allowed you to pass in icon for `ActionSheet.Header`",
            options: Object.keys(icons),
            mapping: icons,
            control: {
                type: "radio",
            },
        },
        primaryAction: {
            control: {
                type: "object",
            },
            description:
                "This prop is meant for `ActionSheet.Footer`. It accepts two property: `Content`, which accepts a string, and `onAction`, which takes a function.",
        },
        secondaryAction: {
            control: {
                type: "object",
            },
            description: "Same as `primaryAction`",
        },
        shouldCloseOnPrimaryButtonClick: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            control: { type: "boolean" },
            description:
                "This prop controls if Action Sheet should be closed or not when primary button was clicked. Default value: true",
        },
        shouldCloseOnSecondaryButtonClick: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            control: { type: "boolean" },
            description:
                "This prop controls if Action Sheet should be closed or not when secondary button was clicked. Default value: true",
        },
        alignment: {
            control: {
                type: "radio",
            },
            options: ["vertical", "horizontal"],
            description:
                "This prop is for `ActionSheet.Footer` buttons alignment",
        },
    },
} satisfies Meta<typeof ActionSheetExampleControlled>;

export default meta;
type Story = StoryObj<typeof ActionSheetExampleControlled>;

export const Controlled: Story = {
    args: {
        expandable: true,
        position: "right",
        primaryAction: {
            content: "Primary",
            onAction: () => null,
        },
        secondaryAction: {
            content: "Secondary",
            onAction: () => null,
        },
        alignment: "vertical",
    },
};
