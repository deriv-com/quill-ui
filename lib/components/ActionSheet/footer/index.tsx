import { FooterProps } from "../types";
import { useContext } from "react";
import { ActionSheetContext } from "../root";
import clsx from "clsx";
import "./footer.scss";
import { Button } from "@components/Button";

const Footer = ({
    primaryAction,
    secondaryAction,
    alignment,
    className,
    shouldCloseOnPrimaryButtonClick = true,
    shouldCloseOnSecondaryButtonClick = true,
    isPrimaryButtonDisabled,
    isSecondaryButtonDisabled,
    ...rest
}: FooterProps) => {
    const { handleClose } = useContext(ActionSheetContext);
    if (!primaryAction && !secondaryAction) return null;

    const primaryActionHandler = () => {
        primaryAction?.onAction();
        if (shouldCloseOnPrimaryButtonClick) handleClose?.();
    };

    const secondaryActionHandler = () => {
        secondaryAction?.onAction();
        if (shouldCloseOnSecondaryButtonClick) handleClose?.();
    };

    return (
        <div
            className={clsx(
                "quill-action-sheet--footer",
                `quill-action-sheet--footer__variant--${alignment}`,
                className,
            )}
            {...rest}
        >
            {primaryAction && (
                <Button
                    onClick={primaryActionHandler}
                    color="black-white"
                    size="lg"
                    label={primaryAction.content}
                    fullWidth
                    disabled={isPrimaryButtonDisabled}
                />
            )}
            {secondaryAction && (
                <Button
                    onClick={secondaryActionHandler}
                    variant="secondary"
                    color="black-white"
                    size="lg"
                    label={secondaryAction.content}
                    fullWidth
                    disabled={isSecondaryButtonDisabled}
                />
            )}
        </div>
    );
};

export default Footer;
