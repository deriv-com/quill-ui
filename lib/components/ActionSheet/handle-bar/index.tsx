import { ComponentProps } from "react";
import "./handle-bar.scss";
import clsx from "clsx";

export interface BarProps extends ComponentProps<"div"> {
    position?: "sticky" | "absolute";
}

const HandleBar = ({ position = "sticky", ...rest }: BarProps) => {
    return (
        <div
            className={clsx(
                "quill-action-sheet--handle-bar",
                `quill-action-sheet--handle-bar--${position}`,
            )}
            data-testid="dt-actionsheet-handle-bar"
            {...rest}
        >
            <span className="quill-action-sheet--handle-bar--line" />
        </div>
    );
};

export default HandleBar;
