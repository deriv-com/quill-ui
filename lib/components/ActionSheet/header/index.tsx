import { ComponentPropsWithoutRef, ReactNode, useContext } from "react";
import { ActionSheetContext } from "../root";
import clsx from "clsx";
import "./header.scss";
import { Heading, Text } from "../../Typography";

interface HeaderProps extends ComponentPropsWithoutRef<"div"> {
    title?: string;
    description?: string;
    icon?: ReactNode;
    closeIcon?: ReactNode;
}

const Header = ({
    className,
    title,
    description,
    icon: Icon,
    closeIcon: CloseIcon,
    ...rest
}: HeaderProps) => {
    const { expandable } = useContext(ActionSheetContext);
    const { handleClose } = useContext(ActionSheetContext);

    return (
        <div
            className={clsx(
                "quill-action-sheet--header",
                `quill-action-sheet--header__expandable--${expandable}`,
                className,
            )}
            data-testid="action-sheet-header"
            {...rest}
        >
            <div className="quill-action-sheet--title">
                <div className="quill-action-sheet--title--icon">{Icon}</div>
                <Heading.H5>{title}</Heading.H5>
                <button
                    className="quill-action-sheet--title--icon quill-action-sheet--title--icon--close"
                    onClick={handleClose}
                >
                    {CloseIcon}
                </button>
            </div>
            <Text>{description}</Text>
        </div>
    );
};

export default Header;
