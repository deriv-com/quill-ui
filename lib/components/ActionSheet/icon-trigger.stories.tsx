import type { Meta, StoryObj } from "@storybook/react";

import { ActionSheetExampleWithIconTrigger } from "./mocks/example";
import { fn } from "@storybook/test";
import {
    LabelPairedPlaceholderCaptionBoldIcon,
    StandaloneXmarkRegularIcon,
} from "@deriv/quill-icons";

const meta: Meta = {
    title: "Components/Action Sheet/Icon Trigger",
    component: ActionSheetExampleWithIconTrigger,
    tags: ["autodocs"],
    parameters: {
        controls: { hideNoControlsWarning: true },
    },
    args: {
        onClose: fn(),
        title: "Title",
        description: "Description",
        closeIcon: <StandaloneXmarkRegularIcon />,
        icon: <LabelPairedPlaceholderCaptionBoldIcon />,
        onOpen: fn(),
        isOpen: false,
    },
    argTypes: {
        isOpen: { table: { disable: true } },
        show: { table: { disable: true } },
        handleOpen: { table: { disable: true } },
        handleClose: { table: { disable: true } },
        onOpen: {
            description:
                "Pass your callback function using this method. It will be triggered on the open function.",
        },
        onClose: {
            description:
                "Pass your callback function using this method. It will be triggered on the open function.",
        },
        expandable: {
            control: { type: "boolean" },
            description:
                "This prop controls the expandability of the bottom sheet.",
        },
        type: {
            options: ["modal", "non-modal"],
            control: { type: "radio" },
            description:
                "This property is used to specify the type, which offers two options: `modal` and `non-modal`. When set to `modal`, it adds an overlay over the entire body, and clicking on it will close the action sheet. On the other hand, when set to `non-modal`, the area outside of the action sheet remains interactive and no overlay.",
        },
        position: {
            options: ["left", "right"],
            control: { type: "radio" },
            description: "This prop will make bottom sheet expandable",
        },
        primaryAction: {
            control: {
                type: "object",
            },
            description:
                "This prop is meant for `ActionSheet.Footer`. It accepts two property: `Content`, which accepts a string, and `onAction`, which takes a function.",
        },
        secondaryAction: {
            control: {
                type: "object",
            },
            description: "Same as `primaryAction`",
        },
        alignment: {
            control: {
                type: "radio",
            },
            options: ["vertical", "horizontal"],
            description:
                "This prop is for `ActionSheet.Footer` buttons alignment",
        },
    },
} satisfies Meta<typeof ActionSheetExampleWithIconTrigger>;

export default meta;
type Story = StoryObj<typeof ActionSheetExampleWithIconTrigger>;

export const IconTrigger: Story = {
    args: {
        expandable: true,
        type: "modal",
        position: "right",
        primaryAction: {
            content: "Primary",
            onAction: () => null,
        },
        secondaryAction: {
            content: "Secondary",
            onAction: () => null,
        },
        alignment: "vertical",
    },
};
