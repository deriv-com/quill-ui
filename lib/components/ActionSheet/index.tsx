import { ReactNode } from "react";
import ActionSheetRoot from "./root";
import ActionSheetHeader from "./header";
import ActionSheetContent from "./content";
import ActionSheetFooter from "./footer";
import ActionSheetHandleBar from "./handle-bar";
import ActionSheetTrigger from "./trigger";
import ActionSheetPortal from "./portal";

type ActionSheetType = {
    Root: typeof ActionSheetRoot;
    Header: typeof ActionSheetHeader;
    Content: typeof ActionSheetContent;
    Footer: typeof ActionSheetFooter;
    HandleBar: typeof ActionSheetHandleBar;
    Trigger: typeof ActionSheetTrigger;
    Portal: typeof ActionSheetPortal;
};

export const ActionSheet: ActionSheetType = ({
    children,
}: {
    children: ReactNode;
}) => {
    return <>{children}</>;
};

ActionSheet.Root = ActionSheetRoot;
ActionSheet.Header = ActionSheetHeader;
ActionSheet.Content = ActionSheetContent;
ActionSheet.Footer = ActionSheetFooter;
ActionSheet.HandleBar = ActionSheetHandleBar;
ActionSheet.Trigger = ActionSheetTrigger;
ActionSheet.Portal = ActionSheetPortal;

export default ActionSheet;
