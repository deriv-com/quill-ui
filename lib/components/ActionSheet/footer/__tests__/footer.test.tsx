import { screen, render } from "@testing-library/react";
import ActionSheet from "@components/ActionSheet";
import userEvent from "@testing-library/user-event";
import { FooterAlignment } from "../../types";

describe("<ActionSheet.Footer/>", () => {
    it("should not render footer if no action passed", () => {
        render(<ActionSheet.Footer aria-label="Footer" />);
        const footerEl = screen.queryByLabelText("Footer");
        expect(footerEl).not.toBeInTheDocument();
    });
    it("should render correctly with className", () => {
        const onActionButton = jest.fn();
        render(
            <ActionSheet.Footer
                className="px-50"
                aria-label="Footer"
                primaryAction={{
                    content: "Primary action",
                    onAction: onActionButton,
                }}
            />,
        );
        const footerEl = screen.getByLabelText("Footer");
        expect(footerEl).toHaveClass("px-50");
    });
    it("should render with primaryAction correctly", async () => {
        const onActionButton = jest.fn();
        render(
            <ActionSheet.Footer
                primaryAction={{
                    content: "Primary action",
                    onAction: onActionButton,
                }}
            />,
        );
        const primaryBtn = screen.getByRole("button", {
            name: "Primary action",
        });
        expect(primaryBtn).toBeInTheDocument();
        await userEvent.click(primaryBtn);
        expect(onActionButton).toHaveBeenCalled();
    });
    it("should render secondaryAction correctly", async () => {
        const onSecondaryAction = jest.fn();
        render(
            <ActionSheet.Footer
                secondaryAction={{
                    content: "Secondary action",
                    onAction: onSecondaryAction,
                }}
            />,
        );
        const secondaryBtn = screen.getByRole("button", {
            name: "Secondary action",
        });
        expect(secondaryBtn).toBeInTheDocument();
        await userEvent.click(secondaryBtn);
        expect(onSecondaryAction).toHaveBeenCalled();
    });
    const alignments: FooterAlignment[] = ["vertical", "horizontal"];

    alignments.forEach((alignment) => {
        it(`should render correctly with ${alignment}`, () => {
            render(
                <ActionSheet.Footer
                    alignment={alignment}
                    aria-label={`${alignment} footer`}
                    primaryAction={{ content: "action", onAction: () => null }}
                />,
            );
            const footer = screen.getByLabelText(`${alignment} footer`);
            expect(footer).toMatchSnapshot();
        });
    });
    it("should not close the component if user clicked on primary button and shouldCloseOnPrimaryButtonClick === false", async () => {
        const onActionButton = jest.fn();
        render(
            <ActionSheet.Footer
                primaryAction={{
                    content: "Primary action",
                    onAction: onActionButton,
                }}
                shouldCloseOnPrimaryButtonClick={false}
            />,
        );
        const primaryBtn = screen.getByRole("button", {
            name: "Primary action",
        });
        expect(primaryBtn).toBeInTheDocument();
        await userEvent.click(primaryBtn);
        expect(onActionButton).toHaveBeenCalled();
        expect(primaryBtn).toBeInTheDocument();
    });
    it("should disable button and do not call a function onAction if isPrimaryButtonDisabled === true", async () => {
        const onActionButton = jest.fn();
        render(
            <ActionSheet.Footer
                primaryAction={{
                    content: "Primary action",
                    onAction: onActionButton,
                }}
                isPrimaryButtonDisabled
            />,
        );
        const primaryBtn = screen.getByRole("button", {
            name: "Primary action",
        });
        await userEvent.click(primaryBtn);
        expect(onActionButton).not.toHaveBeenCalled();
        expect(primaryBtn).toBeDisabled();
    });
});
