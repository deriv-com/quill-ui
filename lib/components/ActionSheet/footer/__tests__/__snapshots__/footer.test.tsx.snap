// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionSheet.Footer/> should render correctly with horizontal 1`] = `
<div
  aria-label="horizontal footer"
  class="quill-action-sheet--footer quill-action-sheet--footer__variant--horizontal"
>
  <button
    class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
    data-state=""
  >
    <span
      class="quill-button-label"
    >
      <p
        class="quill-typography quill-typography__body-text__size--md__weight--bold__decoration--default black"
      >
        action
      </p>
    </span>
  </button>
</div>
`;

exports[`<ActionSheet.Footer/> should render correctly with vertical 1`] = `
<div
  aria-label="vertical footer"
  class="quill-action-sheet--footer quill-action-sheet--footer__variant--vertical"
>
  <button
    class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
    data-state=""
  >
    <span
      class="quill-button-label"
    >
      <p
        class="quill-typography quill-typography__body-text__size--md__weight--bold__decoration--default black"
      >
        action
      </p>
    </span>
  </button>
</div>
`;
